File: src/main/java/systems/conduit/main/core/datastore/backend/MemoryBackend.java
package systems.conduit.main.core.datastore.backend;

import systems.conduit.main.core.datastore.Datastore;
import systems.conduit.main.core.datastore.schema.Schema;
import systems.conduit.main.core.datastore.schema.utils.DatastoreUtils;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author Innectic
 * @since 1/1/2021
 */
public class MemoryBackend implements Datastore {

    private Map<String, Map<Integer, Schema>> internal;

    @Override
    public void attach(Map<String, Object> meta) {
        internal = new HashMap<>();
    }

    @Override
    public void detach() {
        internal.clear();
    }

    @Override
    public void attachSchema(Class<? extends Schema> schema) {
        String name = DatastoreUtils.getNameOfSchema(schema);
        if (!internal.containsKey(name)) internal.put(name, new HashMap<>());
    }

    @Override
    public void insert(Schema schema) {
        String name = DatastoreUtils.getNameOfSchema(schema.getClass());
        if (!internal.containsKey(name)) return;

        internal.get(name).put(schema.getId(), schema);
    }

    @Override
    public void delete(Class<? extends Schema> schema, String key, Object value) {
        Map<String, Object> filter = new HashMap<>();
        filter.put(key, value);

        filterAndDelete(schema, filter);
    }

    @Override
    public void filterAndDelete(Class<? extends Schema> schema, Map<String, Object> filter) {
        String name = DatastoreUtils.getNameOfSchema(schema);
        if (!internal.containsKey(name)) return;
    }

    @Override
    public List<Schema> filter(Class<? extends Schema> schema, Map<String, Object> filter) {
        return null;
    }
}

File: src/main/kotlin/systems/conduit/core/datastore/backend/MemoryBackend.kt
package systems.conduit.core.datastore.backend

import systems.conduit.core.datastore.DatastoreHandler
import systems.conduit.core.datastore.Storable
import java.util.*

/**
 * Datastore backend that puts everything in memory.
 *
 * @author Innectic
 * @since 12/30/2019
 */
class MemoryBackend: DatastoreHandler {

    private var storage: MutableMap<String, Any> = HashMap()

    override fun attach(meta: Map<String, Any>) {
        storage = HashMap()
    }

    override fun detach() {
        storage = hashMapOf()
    }

    override fun set(key: String, value: String) {
        storage[key] = value
    }

    override fun set(key: String, value: Int) {
        storage[key] = value
    }

    override fun set(key: String, value: Float) {
        storage[key] = value
    }

    override fun set(key: String, value: Double) {
        storage[key] = value
    }

    override fun set(key: String, value: Storable<*>) {
        storage[key] = value
    }

    override fun getInt(key: String): Int? {
        val `object` = storage[key] ?: return null
        return try {
            `object` as Int
        } catch (e: ClassCastException) {
            null
        }
    }

    override fun getFloat(key: String): Float? {
        val `object` = storage[key] ?: return null
        return try {
            `object` as Float
        } catch (e: ClassCastException) {
            null
        }
    }

    override fun getDouble(key: String): Double? {
        val `object` = storage[key] ?: return null
        return try {
            `object` as Double
        } catch (e: ClassCastException) {
            null
        }
    }

    override fun getString(key: String): String? {
        val `object` = storage[key] ?: return null
        return try {
            `object` as String
        } catch (e: ClassCastException) {
            null
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T> getCustom(key: String): Storable<T>? {
        val `object` = (storage[key] ?: return null) as? Storable<*> ?: return null
        return try {
            `object` as Storable<T>
        } catch (e: ClassCastException) {
            null
        }
    }

    override fun delete(key: String) {
        storage.remove(key)
    }
}
